File: test/adaptive-viewport/validate-element-grouping-test.py
#!/usr/bin/env python3

"""
Validation script for Element Grouping Preservation Property Test
Verifies that the test file is properly structured and can be executed
"""

import os
import sys

print('=== Validating Element Grouping Preservation Property Test ===\n')

test_file_path = 'test/adaptive-viewport/element-grouping-preservation-property.test.js'
html_file_path = 'test/adaptive-viewport/test-element-grouping-preservation-property.html'

validations_passed = 0
validations_failed = 0

def validate(description, condition):
    global validations_passed, validations_failed
    if condition:
        print(f'✓ {description}')
        validations_passed += 1
        return True
    else:
        print(f'✗ {description}')
        validations_failed += 1
        return False

# Check test file exists
validate('Test file exists', os.path.exists(test_file_path))

# Check HTML runner exists
validate('HTML test runner exists', os.path.exists(html_file_path))

if os.path.exists(test_file_path):
    with open(test_file_path, 'r', encoding='utf-8') as f:
        test_content = f.read()
    
    # Validate test structure
    validate(
        'Test file contains property test function',
        'runElementGroupingPreservationPropertyTest' in test_content
    )
    
    validate(
        'Test validates Requirements 2.2',
        '**Validates: Requirements 2.2**' in test_content
    )
    
    validate(
        'Test is labeled as Property 5',
        'Property 5: Element Grouping Preservation' in test_content
    )
    
    validate(
        'Test uses fast-check library',
        'fc.assert' in test_content and 'fc.asyncProperty' in test_content
    )
    
    validate(
        'Test runs 100 iterations',
        'numRuns: 100' in test_content
    )
    
    # Check for required properties
    validate(
        'Property 1: Grouped elements remain adjacent',
        'Property 1: Grouped elements remain adjacent' in test_content
    )
    
    validate(
        'Property 2: Multiple groups maintain adjacency',
        'Property 2: Multiple groups each maintain internal adjacency' in test_content
    )
    
    validate(
        'Property 3: Group spread remains reasonable',
        'Property 3: Group spread remains reasonable' in test_content
    )
    
    validate(
        'Property 4: Element order preserved',
        'Property 4: Element order within group is preserved' in test_content
    )
    
    # Check for helper functions
    validate(
        'Contains adjacency checking logic',
        'areElementsAdjacent' in test_content and 'isGroupAdjacent' in test_content
    )
    
    validate(
        'Contains group spread calculation',
        'calculateGroupSpread' in test_content
    )
    
    validate(
        'Creates test elements with group IDs',
        'dataset.group' in test_content
    )
    
    validate(
        'Uses LayoutOptimizer',
        'new LayoutOptimizer' in test_content
    )
    
    validate(
        'Calculates optimal layout',
        'calculateOptimalLayout' in test_content
    )
    
    validate(
        'Validates element positions',
        'elementPositions.get' in test_content
    )
    
    validate(
        'Cleans up DOM elements',
        'document.body.removeChild' in test_content
    )
    
    validate(
        'Exports test function',
        'module.exports' in test_content
    )
    
    validate(
        'Returns test results',
        'return results' in test_content
    )
    
    # Check test timeout
    validate(
        'Test has appropriate timeout (40s)',
        'timeout: 40000' in test_content
    )

if os.path.exists(html_file_path):
    with open(html_file_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # Validate HTML structure
    validate(
        'HTML includes test title',
        'Element Grouping Preservation' in html_content
    )
    
    validate(
        'HTML loads fast-check',
        'setup-fast-check.js' in html_content
    )
    
    validate(
        'HTML loads LayoutOptimizer',
        'layout-optimizer.js' in html_content
    )
    
    validate(
        'HTML loads VisibilityDetector',
        'visibility-detector.js' in html_content
    )
    
    validate(
        'HTML loads test file',
        'element-grouping-preservation-property.test.js' in html_content
    )
    
    validate(
        'HTML has run button',
        'runTest()' in html_content
    )
    
    validate(
        'HTML displays test description',
        'logically grouped UI elements' in html_content
    )
    
    validate(
        'HTML lists all 4 properties',
        'Property 1:' in html_content and 
        'Property 2:' in html_content and
        'Property 3:' in html_content and
        'Property 4:' in html_content
    )

# Check dependencies exist
dependencies = [
    'js/adaptive-viewport/layout-optimizer.js',
    'js/adaptive-viewport/visibility-detector.js',
    'js/adaptive-viewport/constants.js',
    'js/adaptive-viewport/types.js',
    'test/adaptive-viewport/setup-fast-check.js'
]

for dep in dependencies:
    validate(f'Dependency exists: {dep}', os.path.exists(dep))

# Print summary
print('\n=== Validation Summary ===')
print(f'Passed: {validations_passed}')
print(f'Failed: {validations_failed}')
print(f'Total: {validations_passed + validations_failed}')

if validations_failed == 0:
    print('\n✅ All validations passed! Test is ready to run.')
    print('\nTo run the test:')
    print('  1. Open test/adaptive-viewport/test-element-grouping-preservation-property.html in a browser')
    print('  2. Click "Run Property Tests"')
    print('  3. Wait for all 4 properties to be tested (100 iterations each)')
    sys.exit(0)
else:
    print('\n❌ Some validations failed. Please review the errors above.')
    sys.exit(1)

File: test/adaptive-viewport/validate-element-grouping-test.js
#!/usr/bin/env node

/**
 * Validation script for Element Grouping Preservation Property Test
 * Verifies that the test file is properly structured and can be executed
 */

const fs = require('fs');
const path = require('path');

console.log('=== Validating Element Grouping Preservation Property Test ===\n');

const testFilePath = path.join(__dirname, 'element-grouping-preservation-property.test.js');
const htmlFilePath = path.join(__dirname, 'test-element-grouping-preservation-property.html');

let validationsPassed = 0;
let validationsFailed = 0;

function validate(description, condition) {
  if (condition) {
    console.log(`✓ ${description}`);
    validationsPassed++;
    return true;
  } else {
    console.error(`✗ ${description}`);
    validationsFailed++;
    return false;
  }
}

// Check test file exists
validate('Test file exists', fs.existsSync(testFilePath));

// Check HTML runner exists
validate('HTML test runner exists', fs.existsSync(htmlFilePath));

if (fs.existsSync(testFilePath)) {
  const testContent = fs.readFileSync(testFilePath, 'utf8');
  
  // Validate test structure
  validate(
    'Test file contains property test function',
    testContent.includes('runElementGroupingPreservationPropertyTest')
  );
  
  validate(
    'Test validates Requirements 2.2',
    testContent.includes('**Validates: Requirements 2.2**')
  );
  
  validate(
    'Test is labeled as Property 5',
    testContent.includes('Property 5: Element Grouping Preservation')
  );
  
  validate(
    'Test uses fast-check library',
    testContent.includes('fc.assert') && testContent.includes('fc.asyncProperty')
  );
  
  validate(
    'Test runs 100 iterations',
    testContent.includes('numRuns: 100')
  );
  
  // Check for required properties
  validate(
    'Property 1: Grouped elements remain adjacent',
    testContent.includes('Property 1: Grouped elements remain adjacent')
  );
  
  validate(
    'Property 2: Multiple groups maintain adjacency',
    testContent.includes('Property 2: Multiple groups each maintain internal adjacency')
  );
  
  validate(
    'Property 3: Group spread remains reasonable',
    testContent.includes('Property 3: Group spread remains reasonable')
  );
  
  validate(
    'Property 4: Element order preserved',
    testContent.includes('Property 4: Element order within group is preserved')
  );
  
  // Check for helper functions
  validate(
    'Contains adjacency checking logic',
    testContent.includes('areElementsAdjacent') && testContent.includes('isGroupAdjacent')
  );
  
  validate(
    'Contains group spread calculation',
    testContent.includes('calculateGroupSpread')
  );
  
  validate(
    'Creates test elements with group IDs',
    testContent.includes('dataset.group')
  );
  
  validate(
    'Uses LayoutOptimizer',
    testContent.includes('new LayoutOptimizer')
  );
  
  validate(
    'Calculates optimal layout',
    testContent.includes('calculateOptimalLayout')
  );
  
  validate(
    'Validates element positions',
    testContent.includes('elementPositions.get')
  );
  
  validate(
    'Cleans up DOM elements',
    testContent.includes('document.body.removeChild')
  );
  
  validate(
    'Exports test function',
    testContent.includes('module.exports')
  );
  
  validate(
    'Returns test results',
    testContent.includes('return results')
  );
  
  // Check test timeout
  validate(
    'Test has appropriate timeout (40s)',
    testContent.includes('timeout: 40000')
  );
}

if (fs.existsSync(htmlFilePath)) {
  const htmlContent = fs.readFileSync(htmlFilePath, 'utf8');
  
  // Validate HTML structure
  validate(
    'HTML includes test title',
    htmlContent.includes('Element Grouping Preservation')
  );
  
  validate(
    'HTML loads fast-check',
    htmlContent.includes('setup-fast-check.js')
  );
  
  validate(
    'HTML loads LayoutOptimizer',
    htmlContent.includes('layout-optimizer.js')
  );
  
  validate(
    'HTML loads VisibilityDetector',
    htmlContent.includes('visibility-detector.js')
  );
  
  validate(
    'HTML loads test file',
    htmlContent.includes('element-grouping-preservation-property.test.js')
  );
  
  validate(
    'HTML has run button',
    htmlContent.includes('runTest()')
  );
  
  validate(
    'HTML displays test description',
    htmlContent.includes('logically grouped UI elements')
  );
  
  validate(
    'HTML lists all 4 properties',
    htmlContent.includes('Property 1:') && 
    htmlContent.includes('Property 2:') &&
    htmlContent.includes('Property 3:') &&
    htmlContent.includes('Property 4:')
  );
}

// Check dependencies exist
const dependencies = [
  'js/adaptive-viewport/layout-optimizer.js',
  'js/adaptive-viewport/visibility-detector.js',
  'js/adaptive-viewport/constants.js',
  'js/adaptive-viewport/types.js',
  'test/adaptive-viewport/setup-fast-check.js'
];

dependencies.forEach(dep => {
  const depPath = path.join(__dirname, '..', '..', dep);
  validate(`Dependency exists: ${dep}`, fs.existsSync(depPath));
});

// Print summary
console.log('\n=== Validation Summary ===');
console.log(`Passed: ${validationsPassed}`);
console.log(`Failed: ${validationsFailed}`);
console.log(`Total: ${validationsPassed + validationsFailed}`);

if (validationsFailed === 0) {
  console.log('\n✅ All validations passed! Test is ready to run.');
  console.log('\nTo run the test:');
  console.log('  1. Open test/adaptive-viewport/test-element-grouping-preservation-property.html in a browser');
  console.log('  2. Click "Run Property Tests"');
  console.log('  3. Wait for all 4 properties to be tested (100 iterations each)');
  process.exit(0);
} else {
  console.error('\n❌ Some validations failed. Please review the errors above.');
  process.exit(1);
}
